File: Sources/CpdbShared/Capture/ContentIdentity.swift
import Foundation
import CryptoKit

/// Canonical content identity, v2 (`idv2-r1`).
///
/// Entry identity = SHA-256 of the **primary** content only, chosen by a rung
/// chain (image → file → url → normalized text → color → full-set fallback).
/// Identity matches user-perceived content so two devices that copy the same
/// thing converge with no coordination; storage keeps full capture fidelity.
///
/// THIS FILE IS A PORT OF THE REFERENCE IMPLEMENTATION `Tools/gen_hash_vectors.py`.
/// The authoritative test vectors live in `Tests/Fixtures/hash-vectors-v2.json`
/// and `HashVectors.swift` asserts every one. The Windows port
/// (`CpdbWin.Core.ContentIdentity`) must reproduce the same hexes. The prose
/// spec is `docs/canonical-hash-v2.md` §2; where code and prose disagree, the
/// vectors JSON wins. Any change here is a wire-format change: bump
/// `revision`, regenerate the vectors, update both platforms in lockstep.
public enum ContentIdentity {

    /// Logged with every computed hash; never a hash input.
    public static let revision = "idv2-r1"

    /// The rung that produced an identity. Carried on the CloudKit wire as
    /// `identityTag` and used for deterministic conflict resolution: a lower
    /// `rungIndex` was computed from more-authoritative input, so on an
    /// equal-version hash fork the lower rung wins (missing flavors can only
    /// demote identity *down* the chain). See docs §2.3 / §5.3.
    public enum Tag: String, Sendable {
        case image, file, url, text, color, fallback

        public var rungIndex: Int {
            switch self {
            case .image:    return 1
            case .file:     return 2
            case .url:      return 3
            case .text:     return 4
            case .color:    return 5
            case .fallback: return 6
            }
        }
    }

    // MARK: - Pinned constants (mirror docs §2.3 and Tools/gen_hash_vectors.py exactly)

    static let sharedPasteboardMarker =
        "group.com.apple.coreservices.useractivityd/shared-pasteboard/"

    /// U+0020 space, U+0009 tab, U+000A LF, U+000D CR, U+00A0 NBSP. Exhaustive.
    static let urlTrimSet: Set<Character> = ["\u{0020}", "\u{0009}", "\u{000A}", "\u{000D}", "\u{00A0}"]

    static let imageUTIs = ["public.png", "public.jpeg", "public.tiff",
                            "public.heic", "public.heif", "public.image"]
    static let imageMinBytes = 1024

    static let colorUTIs = ["com.apple.cocoa.pasteboard.color", "public.color"]

    static let textUTF8UTIs = ["public.utf8-plain-text", "public.plain-text"]
    static let textUTF16External = "public.utf16-external-plain-text"
    static let textUTF16 = "public.utf16-plain-text"
    static let textHTMLLastResort = "public.html"

    /// Fallback rung only. Excluded from the v1-style emission.
    static let volatileExact: Set<String> = [
        "public.text",
        "com.apple.is-remote-clipboard",
        "com.apple.traditional-mac-plain-text",
        "org.chromium.source-url",
        "org.chromium.internal.source-rfh-token",
        "com.apple.WebKit.custom-pasteboard-data",
        "com.apple.linkpresentation.metadata",
        "com.apple.icns",
        "com.raycast.RestoredType",
        "com.apple.security.sandbox-extension-dict",
        "public.utf16-plain-text",
        "public.utf16-external-plain-text",
    ]
    static let volatilePrefixes = ["com.apple.iWork.pasteboardState.", "dyn."]

    static func isVolatile(_ uti: String) -> Bool {
        volatileExact.contains(uti) || volatilePrefixes.contains { uti.hasPrefix($0) }
    }

    // MARK: - Public API

    /// Compute identity over a list of pasteboard items (each a flavor list).
    /// Returns the rung tag and the 32-byte SHA-256 `content_hash`.
    public static func compute(items: [[CanonicalHash.Flavor]]) -> (tag: Tag, hash: Data) {
        let flat = flatten(items)
        let (tag, value) = identity(flat)
        var hasher = SHA256()
        hasher.update(data: Data(tag.rawValue.utf8))
        hasher.update(data: Data([0x00]))
        hasher.update(data: value)
        return (tag, Data(hasher.finalize()))
    }

    /// Convenience: single-item snapshot.
    public static func compute(flavors: [CanonicalHash.Flavor]) -> (tag: Tag, hash: Data) {
        compute(items: [flavors])
    }

    // MARK: - Capture policy (§2.4 skip rules — NOT part of the hash contract)

    /// True iff every flavor in the snapshot is zero-length. Such a snapshot
    /// carries no content at all; the Ingestor skips it (belt over the
    /// empty-after-normalization → fallback suspender in the rung chain).
    public static func snapshotIsAllEmpty(items: [[CanonicalHash.Flavor]]) -> Bool {
        let flat = flatten(items)
        return flat.isEmpty || flat.values.allSatisfy { $0.isEmpty }
    }

    /// True iff the snapshot is a Universal Clipboard *file* echo with no
    /// other substantive content: a `public.file-url` pointing into the
    /// useractivityd shared-pasteboard container, and no image ≥ 1 KB, no
    /// non-empty text source, no `public.url`, no color. These are
    /// Mac-to-Mac Finder-copy mirrors — the receiving Mac materializes the
    /// remote file under a fresh UUID per transfer, so the path can never
    /// dedup and the row is pure noise. Echoes that DO carry image bytes or
    /// text (iPhone-origin copies, for which the echo is the only record
    /// anywhere) key via rungs 1/4 and are kept. See docs §2.4.
    public static func isSoleSharedPasteboardEcho(items: [[CanonicalHash.Flavor]]) -> Bool {
        let flat = flatten(items)
        guard let b = flat["public.file-url"], let s = strictUTF8(b),
              s.contains(sharedPasteboardMarker) else { return false }
        // Any other substantive flavor rescues the snapshot.
        for uti in imageUTIs {
            if let img = flat[uti], img.count >= imageMinBytes { return false }
        }
        if let t = bestText(flat), !normalizeText(t).isEmpty { return false }
        if flat["public.url"] != nil { return false }
        for uti in colorUTIs where flat[uti] != nil { return false }
        return true
    }

    // MARK: - Flattening (§2.2 — first occurrence of each UTI wins)

    static func flatten(_ items: [[CanonicalHash.Flavor]]) -> [String: Data] {
        var flat: [String: Data] = [:]
        for item in items {
            for f in item where flat[f.uti] == nil {
                flat[f.uti] = f.data
            }
        }
        return flat
    }

    // MARK: - The rung chain (§2.3)

    static func identity(_ flat: [String: Data]) -> (Tag, Data) {
        // Rung 1 — IMAGE
        for uti in imageUTIs {
            if let b = flat[uti], b.count >= imageMinBytes {
                return (.image, b)
            }
        }

        // Rung 2 — FILE
        if let b = flat["public.file-url"], let s = strictUTF8(b),
           !s.contains(sharedPasteboardMarker) {
            let decoded = stripOneTrailingSlash(percentDecode(s))
            return (.file, Data(decoded.utf8))
        }
        // (decode failure OR shared-pasteboard echo: stored, not identity — fall through)

        // Rung 3 — URL
        if let b = flat["public.url"], let s0 = strictUTF8(b) {
            let s = stripOneTrailingSlash(trimURLSet(s0))
            if !s.isEmpty {
                return (.url, Data(s.utf8))
            }
        }

        // Rung 4 — TEXT (with URL-shaped-text promotion)
        if let raw = bestText(flat) {
            let t = normalizeText(raw)
            if !t.isEmpty {
                let p = trimURLSet(t)
                if looksLikeURLPortable(p) {
                    return (.url, Data(stripOneTrailingSlash(p).utf8))
                }
                return (.text, Data(t.utf8))
            }
            // empty after normalization: fall to fallback
        }

        // Rung 5 — COLOR
        for uti in colorUTIs {
            if let b = flat[uti] {
                return (.color, b)
            }
        }

        // Rung 6 — FALLBACK
        var kept = flat.filter { !isVolatile($0.key) }
        if kept.isEmpty { kept = flat }
        return (.fallback, v1Emission(kept))
    }

    // MARK: - Primitives

    /// Strict UTF-8: nil on any invalid sequence (never substitutes U+FFFD).
    static func strictUTF8(_ b: Data) -> String? {
        String(data: b, encoding: .utf8)
    }

    /// Decode a UTF-16 flavor per §2.3. BOM present → use its endianness and
    /// strip it; BOM-less → little-endian everywhere. Strict: odd length or any
    /// unpaired surrogate → nil. Manual code-unit walk so Swift/C#/Python agree
    /// exactly (we do not rely on platform `String(data:encoding:)` quirks).
    static func decodeUTF16(_ b: Data) -> String? {
        if b.count % 2 != 0 { return nil }
        let bytes = [UInt8](b)
        var idx = 0
        let bigEndian: Bool
        if bytes.count >= 2 && bytes[0] == 0xFF && bytes[1] == 0xFE {
            bigEndian = false; idx = 2
        } else if bytes.count >= 2 && bytes[0] == 0xFE && bytes[1] == 0xFF {
            bigEndian = true; idx = 2
        } else {
            bigEndian = false  // BOM-less default
        }
        var scalars: [Unicode.Scalar] = []
        func unit(_ i: Int) -> UInt16 {
            let lo = UInt16(bytes[i]); let hi = UInt16(bytes[i + 1])
            return bigEndian ? (lo << 8 | hi) : (hi << 8 | lo)
        }
        while idx + 1 < bytes.count {
            let u = unit(idx); idx += 2
            if u >= 0xD800 && u <= 0xDBFF {
                // high surrogate: must be followed by a low surrogate
                guard idx + 1 < bytes.count else { return nil }
                let low = unit(idx)
                guard low >= 0xDC00 && low <= 0xDFFF else { return nil }
                idx += 2
                let code = 0x10000 + (UInt32(u - 0xD800) << 10) + UInt32(low - 0xDC00)
                guard let sc = Unicode.Scalar(code) else { return nil }
                scalars.append(sc)
            } else if u >= 0xDC00 && u <= 0xDFFF {
                return nil  // lone low surrogate
            } else {
                guard let sc = Unicode.Scalar(UInt32(u)) else { return nil }
                scalars.append(sc)
            }
        }
        var s = ""
        s.unicodeScalars.append(contentsOf: scalars)
        return s
    }

    /// Rung-4 decode ladder. A source that fails strict decode is skipped.
    static func bestText(_ flat: [String: Data]) -> String? {
        for uti in textUTF8UTIs {
            if let b = flat[uti], let s = strictUTF8(b) { return s }
        }
        if let b = flat[textUTF16External], let s = decodeUTF16(b) { return s }
        if let b = flat[textUTF16], let s = decodeUTF16(b) { return s }
        if let b = flat[textHTMLLastResort], let s = strictUTF8(b) { return s }
        return nil
    }

    /// Byte-level normalization: strip ONE leading BOM (U+FEFF), CRLF→LF, lone
    /// CR→LF. No outer trim (V5 rejected). No Unicode NFC (divergence risk).
    static func normalizeText(_ input: String) -> String {
        var s = Substring(input)
        if s.first == "\u{FEFF}" { s = s.dropFirst() }
        return s.replacingOccurrences(of: "\r\n", with: "\n")
                .replacingOccurrences(of: "\r", with: "\n")
    }

    static func trimURLSet(_ s: String) -> String {
        String(s.drop(while: { urlTrimSet.contains($0) })
                .reversed().drop(while: { urlTrimSet.contains($0) }).reversed())
    }

    /// Remove exactly one trailing "/" unless the string ends with "://".
    static func stripOneTrailingSlash(_ s: String) -> String {
        if s.hasSuffix("/") && !s.hasSuffix("://") {
            return String(s.dropLast())
        }
        return s
    }

    /// p is already URL-trimmed. Portable lowercase-exact http(s) check for the
    /// text→url promotion. Deliberately narrow so Swift/C# never diverge.
    static func looksLikeURLPortable(_ p: String) -> Bool {
        if p.count > 2048 { return false }
        guard p.hasPrefix("http://") || p.hasPrefix("https://") else { return false }
        if p.contains(where: { urlTrimSet.contains($0) }) { return false }
        guard let schemeRange = p.range(of: "://") else { return false }
        let after = p[schemeRange.upperBound...]
        // authority = up to first of / ? # (or end); must be non-empty
        let authEnd = after.firstIndex(where: { $0 == "/" || $0 == "?" || $0 == "#" }) ?? after.endIndex
        return authEnd > after.startIndex
    }

    /// Percent-decode matching Python's `urllib.parse.unquote(s)` (UTF-8,
    /// errors="replace"): "%" + two ASCII hex → that byte; otherwise the
    /// literal "%" is kept. The collected byte stream is decoded as UTF-8 with
    /// U+FFFD substitution. Identical behaviour on both platforms by
    /// construction (manual, no stdlib percent API).
    static func percentDecode(_ s: String) -> String {
        let src = [UInt8](s.utf8)
        var out: [UInt8] = []
        out.reserveCapacity(src.count)
        var i = 0
        func hexVal(_ c: UInt8) -> UInt8? {
            switch c {
            case 0x30...0x39: return c - 0x30
            case 0x41...0x46: return c - 0x41 + 10
            case 0x61...0x66: return c - 0x61 + 10
            default: return nil
            }
        }
        while i < src.count {
            if src[i] == 0x25, i + 2 < src.count,            // '%'
               let hi = hexVal(src[i + 1]), let lo = hexVal(src[i + 2]) {
                out.append(hi << 4 | lo)
                i += 3
            } else {
                out.append(src[i])
                i += 1
            }
        }
        return String(decoding: out, as: UTF8.self)
    }

    /// v1 single-item canonical emission over the kept flavors, BYTE-WISE
    /// sorted by the UTF-8 encoding of the UTI (NOT Swift `<` collation, NOT
    /// UTF-16 ordinal): for each `uti||0x00||u64be(len)||bytes`, then a single
    /// trailing 0x01.
    static func v1Emission(_ kept: [String: Data]) -> Data {
        let sorted = kept.sorted { Array($0.key.utf8).lexicographicallyPrecedes(Array($1.key.utf8)) }
        var out = Data()
        for (uti, bytes) in sorted {
            out.append(Data(uti.utf8))
            out.append(0x00)
            var lenBE = UInt64(bytes.count).bigEndian
            withUnsafeBytes(of: &lenBE) { out.append(contentsOf: $0) }
            out.append(bytes)
        }
        out.append(0x01)
        return out
    }
}

File: windows/CpdbWin.Core/Capture/ContentIdentity.cs
using System.Security.Cryptography;
using System.Text;

namespace CpdbWin.Core.Capture;

/// <summary>
/// Canonical content identity, v2 (<c>idv2-r1</c>). Windows port of
/// <c>Sources/CpdbShared/Capture/ContentIdentity.swift</c> /
/// <c>Tools/gen_hash_vectors.py</c>.
///
/// <para>
/// Entry identity = SHA-256 of the <b>primary</b> content only, chosen by a
/// rung chain: image → file → url → normalized text → color →
/// full-set fallback. Identity matches user-perceived content so a copy
/// captured on two devices converges with no coordination; volatile sidecar
/// formats (Chromium session tokens, Universal Clipboard re-publication
/// noise, link-preview metadata, encoding variants) jitter byte-for-byte
/// between otherwise-identical copies and the old full-set hash forked one
/// logical clip into many rows.
/// </para>
///
/// <para>
/// THIS FILE IS A PORT OF THE REFERENCE IMPLEMENTATION
/// <c>Tools/gen_hash_vectors.py</c>. The authoritative test vectors live in
/// <c>Tests/Fixtures/hash-vectors-v2.json</c>; <c>ContentIdentityVectorsTests</c>
/// asserts every expect_hex. Prose spec: <c>docs/canonical-hash-v2.md</c> §2.
/// Where code and prose disagree, the vectors JSON wins. Any change here is
/// a wire-format change: bump <see cref="Revision"/>, regenerate the
/// vectors via the Python script, and update Swift in lockstep.
/// </para>
/// </summary>
public static class ContentIdentity
{
    /// <summary>Logged with every computed hash; never a hash input.</summary>
    public const string Revision = "idv2-r1";

    /// <summary>The rung that produced an identity. Stored in
    /// <c>entries.identity_tag</c>.</summary>
    public enum Tag
    {
        Image,
        File,
        Url,
        Text,
        Color,
        Fallback,
    }

    /// <summary>Lowercase canonical tag string (used as a hash input).
    /// Stable wire form — never reformat.</summary>
    public static string TagString(Tag t) => t switch
    {
        Tag.Image    => "image",
        Tag.File     => "file",
        Tag.Url      => "url",
        Tag.Text     => "text",
        Tag.Color    => "color",
        Tag.Fallback => "fallback",
        _ => throw new ArgumentOutOfRangeException(nameof(t), t, null),
    };

    /// <summary>Parse a tag string back to the enum. Returns null for any
    /// unrecognised input (forward-compatible if Mac adds new rungs).</summary>
    public static Tag? ParseTag(string? s) => s switch
    {
        "image"    => Tag.Image,
        "file"     => Tag.File,
        "url"      => Tag.Url,
        "text"     => Tag.Text,
        "color"    => Tag.Color,
        "fallback" => Tag.Fallback,
        _          => null,
    };

    public static int RungIndex(Tag t) => t switch
    {
        Tag.Image    => 1,
        Tag.File     => 2,
        Tag.Url      => 3,
        Tag.Text     => 4,
        Tag.Color    => 5,
        Tag.Fallback => 6,
        _ => throw new ArgumentOutOfRangeException(nameof(t), t, null),
    };

    // ── Pinned constants (mirror docs §2.3 + gen_hash_vectors.py exactly) ──

    public const string SharedPasteboardMarker =
        "group.com.apple.coreservices.useractivityd/shared-pasteboard/";

    /// <summary>U+0020 space, U+0009 tab, U+000A LF, U+000D CR, U+00A0 NBSP.
    /// Exhaustive — no locale, no Unicode "whitespace" class.</summary>
    public static readonly HashSet<char> UrlTrimSet = new()
    {
        (char)0x20, (char)0x09, (char)0x0A, (char)0x0D, (char)0xA0,
    };

    public static readonly string[] ImageUtis =
        { "public.png", "public.jpeg", "public.tiff", "public.heic", "public.heif", "public.image" };

    public const int ImageMinBytes = 1024;

    public static readonly string[] ColorUtis =
        { "com.apple.cocoa.pasteboard.color", "public.color" };

    public static readonly string[] TextUtf8Utis = { "public.utf8-plain-text", "public.plain-text" };
    public const string TextUtf16External = "public.utf16-external-plain-text";
    public const string TextUtf16         = "public.utf16-plain-text";
    public const string TextHtmlLastResort = "public.html";

    /// <summary>Fallback rung only. Excluded from the v1-style emission.
    /// The semantic rungs make this list irrelevant everywhere except
    /// <c>kind=other</c> junk — that inversion is the point: a new
    /// volatile UTI in 2027 needs zero code change.</summary>
    public static readonly HashSet<string> VolatileExact = new(StringComparer.Ordinal)
    {
        "public.text",
        "com.apple.is-remote-clipboard",
        "com.apple.traditional-mac-plain-text",
        "org.chromium.source-url",
        "org.chromium.internal.source-rfh-token",
        "com.apple.WebKit.custom-pasteboard-data",
        "com.apple.linkpresentation.metadata",
        "com.apple.icns",
        "com.raycast.RestoredType",
        "com.apple.security.sandbox-extension-dict",
        "public.utf16-plain-text",
        "public.utf16-external-plain-text",
    };

    public static readonly string[] VolatilePrefixes = { "com.apple.iWork.pasteboardState.", "dyn." };

    public static bool IsVolatile(string uti)
        => VolatileExact.Contains(uti)
        || VolatilePrefixes.Any(p => uti.StartsWith(p, StringComparison.Ordinal));

    // ── Public API ──────────────────────────────────────────────────────

    /// <summary>Compute identity over a list of pasteboard items (each a
    /// flavor list). Returns the rung tag and the 32-byte SHA-256
    /// <c>content_hash</c> the migrator + ingestor + importer all use.</summary>
    public static (Tag Tag, byte[] Hash) Compute(IReadOnlyList<IReadOnlyList<CanonicalHash.Flavor>> items)
    {
        var flat = Flatten(items);
        var (tag, value) = Identity(flat);
        using var h = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        h.AppendData(Encoding.UTF8.GetBytes(TagString(tag)));
        h.AppendData(new byte[] { 0x00 });
        h.AppendData(value);
        return (tag, h.GetHashAndReset());
    }

    /// <summary>Convenience: single-item snapshot.</summary>
    public static (Tag Tag, byte[] Hash) Compute(IReadOnlyList<CanonicalHash.Flavor> flavors)
        => Compute(new[] { flavors });

    // ── Flattening (§2.2 — first occurrence of each UTI wins) ────────────

    internal static Dictionary<string, byte[]> Flatten(
        IReadOnlyList<IReadOnlyList<CanonicalHash.Flavor>> items)
    {
        var flat = new Dictionary<string, byte[]>(StringComparer.Ordinal);
        foreach (var item in items)
            foreach (var f in item)
                if (!flat.ContainsKey(f.Uti))
                    flat[f.Uti] = f.Data.ToArray();
        return flat;
    }

    // ── The rung chain (§2.3) ────────────────────────────────────────────

    internal static (Tag Tag, byte[] Value) Identity(Dictionary<string, byte[]> flat)
    {
        // Rung 1 — IMAGE
        foreach (var uti in ImageUtis)
        {
            if (flat.TryGetValue(uti, out var b) && b.Length >= ImageMinBytes)
                return (Tag.Image, b);
        }

        // Rung 2 — FILE
        if (flat.TryGetValue("public.file-url", out var fb))
        {
            var s = StrictUtf8(fb);
            if (s is not null && !s.Contains(SharedPasteboardMarker, StringComparison.Ordinal))
            {
                var decoded = StripOneTrailingSlash(PercentDecode(s));
                return (Tag.File, Encoding.UTF8.GetBytes(decoded));
            }
            // decode failure OR shared-pasteboard echo: stored, not identity
        }

        // Rung 3 — URL
        if (flat.TryGetValue("public.url", out var ub))
        {
            var s0 = StrictUtf8(ub);
            if (s0 is not null)
            {
                var s = StripOneTrailingSlash(TrimUrlSet(s0));
                if (s.Length > 0)
                    return (Tag.Url, Encoding.UTF8.GetBytes(s));
            }
        }

        // Rung 4 — TEXT (with URL-shaped-text promotion)
        var raw = BestText(flat);
        if (raw is not null)
        {
            var t = NormalizeText(raw);
            if (t.Length > 0)
            {
                var p = TrimUrlSet(t);
                if (LooksLikeUrlPortable(p))
                    return (Tag.Url, Encoding.UTF8.GetBytes(StripOneTrailingSlash(p)));
                return (Tag.Text, Encoding.UTF8.GetBytes(t));
            }
            // empty after normalization: fall to fallback (avoids the
            // sha256("text\0") mega-cluster the audit found)
        }

        // Rung 5 — COLOR
        foreach (var uti in ColorUtis)
        {
            if (flat.TryGetValue(uti, out var cb))
                return (Tag.Color, cb);
        }

        // Rung 6 — FALLBACK
        var kept = flat.Where(kv => !IsVolatile(kv.Key))
                       .Select(kv => new CanonicalHash.Flavor(kv.Key, kv.Value))
                       .ToList();
        if (kept.Count == 0)
            kept = flat.Select(kv => new CanonicalHash.Flavor(kv.Key, kv.Value)).ToList();
        // Use the RAW v1 emission bytes; the outer Compute() above wraps
        // them with one SHA-256 over "fallback" || 0x00 || EmitV1(...).
        // Calling CanonicalHash.Compute here would double-hash.
        return (Tag.Fallback, CanonicalHash.EmitV1(new[] { (IReadOnlyList<CanonicalHash.Flavor>)kept }));
    }

    // ── Primitives (mirror gen_hash_vectors.py one-for-one) ──────────────

    /// <summary>Strict UTF-8: null on any invalid sequence (never substitutes
    /// U+FFFD).</summary>
    internal static string? StrictUtf8(byte[] bytes)
    {
        try
        {
            var enc = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
            return enc.GetString(bytes);
        }
        catch (DecoderFallbackException)
        {
            return null;
        }
    }

    /// <summary>Decode a UTF-16 flavor per §2.3. BOM present → use its
    /// endianness and strip it; BOM-less → little-endian everywhere.
    /// Strict: odd length or any unpaired surrogate → null. Manual
    /// code-unit walk so Swift/C#/Python agree exactly.</summary>
    internal static string? DecodeUtf16(byte[] bytes)
    {
        if (bytes.Length % 2 != 0) return null;
        bool bigEndian;
        int idx;
        if (bytes.Length >= 2 && bytes[0] == 0xFF && bytes[1] == 0xFE)
        { bigEndian = false; idx = 2; }
        else if (bytes.Length >= 2 && bytes[0] == 0xFE && bytes[1] == 0xFF)
        { bigEndian = true;  idx = 2; }
        else
        { bigEndian = false; idx = 0; }  // BOM-less default

        var sb = new StringBuilder((bytes.Length - idx) / 2);

        ushort Unit(int i)
        {
            byte lo = bytes[i], hi = bytes[i + 1];
            return bigEndian ? (ushort)((lo << 8) | hi) : (ushort)((hi << 8) | lo);
        }

        while (idx + 1 < bytes.Length)
        {
            ushort u = Unit(idx); idx += 2;
            if (u >= 0xD800 && u <= 0xDBFF)
            {
                // high surrogate: must be followed by a low surrogate
                if (idx + 1 >= bytes.Length) return null;
                ushort low = Unit(idx);
                if (low < 0xDC00 || low > 0xDFFF) return null;
                idx += 2;
                int code = 0x10000 + ((u - 0xD800) << 10) + (low - 0xDC00);
                sb.Append(char.ConvertFromUtf32(code));
            }
            else if (u >= 0xDC00 && u <= 0xDFFF)
            {
                return null;  // lone low surrogate
            }
            else
            {
                sb.Append((char)u);
            }
        }
        return sb.ToString();
    }

    /// <summary>Rung-4 decode ladder. A source that fails strict decode is
    /// skipped (not substituted).</summary>
    internal static string? BestText(Dictionary<string, byte[]> flat)
    {
        foreach (var uti in TextUtf8Utis)
        {
            if (flat.TryGetValue(uti, out var b))
            {
                var s = StrictUtf8(b);
                if (s is not null) return s;
            }
        }
        if (flat.TryGetValue(TextUtf16External, out var u16e))
        {
            var s = DecodeUtf16(u16e);
            if (s is not null) return s;
        }
        if (flat.TryGetValue(TextUtf16, out var u16))
        {
            var s = DecodeUtf16(u16);
            if (s is not null) return s;
        }
        if (flat.TryGetValue(TextHtmlLastResort, out var html))
        {
            var s = StrictUtf8(html);
            if (s is not null) return s;
        }
        return null;
    }

    /// <summary>Byte-level normalization after strict validation: strip ONE
    /// leading BOM (U+FEFF), CRLF → LF, lone CR → LF. No outer trim. No
    /// Unicode NFC (platform-divergence risk).</summary>
    internal static string NormalizeText(string s)
    {
        if (s.Length > 0 && s[0] == (char)0xFEFF) s = s.Substring(1);
        s = s.Replace("\r\n", "\n").Replace("\r", "\n");
        return s;
    }

    /// <summary>Strip leading/trailing codepoints in
    /// <see cref="UrlTrimSet"/> only.</summary>
    internal static string TrimUrlSet(string s)
    {
        int i = 0, j = s.Length;
        while (i < j && UrlTrimSet.Contains(s[i])) i++;
        while (j > i && UrlTrimSet.Contains(s[j - 1])) j--;
        return s.Substring(i, j - i);
    }

    /// <summary>Remove exactly one trailing '/' unless the string ends with
    /// "://".</summary>
    internal static string StripOneTrailingSlash(string s)
    {
        if (s.EndsWith("/", StringComparison.Ordinal) && !s.EndsWith("://", StringComparison.Ordinal))
            return s.Substring(0, s.Length - 1);
        return s;
    }

    /// <summary><paramref name="p"/> is already URL-trimmed. Portable
    /// lowercase-exact http(s) check used for the text→url promotion.
    /// Deliberately narrow so Swift/C#/Python never diverge.</summary>
    internal static bool LooksLikeUrlPortable(string p)
    {
        if (p.Length > 2048) return false;
        if (!(p.StartsWith("http://",  StringComparison.Ordinal)
           || p.StartsWith("https://", StringComparison.Ordinal))) return false;
        for (int i = 0; i < p.Length; i++)
            if (UrlTrimSet.Contains(p[i])) return false;
        int scheme = p.IndexOf("://", StringComparison.Ordinal);
        if (scheme < 0) return false;
        int afterStart = scheme + 3;
        // authority = up to first of / ? # (or end); must be non-empty.
        int cut = p.Length;
        foreach (var sep in new[] { '/', '?', '#' })
        {
            int k = p.IndexOf(sep, afterStart);
            if (k >= 0 && k < cut) cut = k;
        }
        return cut > afterStart;
    }

    /// <summary>Percent-decode matching Python's
    /// <c>urllib.parse.unquote(s, errors="replace")</c>: "%" + two ASCII
    /// hex → that byte; otherwise the literal "%" is kept. The collected
    /// byte stream is decoded as UTF-8 with U+FFFD substitution on bad
    /// bytes (the default <see cref="Encoding.UTF8"/> already does this).
    /// Manual implementation, no stdlib percent API, identical on every
    /// platform by construction.</summary>
    internal static string PercentDecode(string s)
    {
        var src = Encoding.UTF8.GetBytes(s);
        var outBytes = new List<byte>(src.Length);
        int i = 0;
        while (i < src.Length)
        {
            if (src[i] == (byte)'%' && i + 2 < src.Length)
            {
                int hi = HexVal(src[i + 1]);
                int lo = HexVal(src[i + 2]);
                if (hi >= 0 && lo >= 0)
                {
                    outBytes.Add((byte)((hi << 4) | lo));
                    i += 3;
                    continue;
                }
            }
            outBytes.Add(src[i]);
            i++;
        }
        return Encoding.UTF8.GetString(outBytes.ToArray());
    }

    private static int HexVal(byte b) => b switch
    {
        >= (byte)'0' and <= (byte)'9' => b - (byte)'0',
        >= (byte)'A' and <= (byte)'F' => b - (byte)'A' + 10,
        >= (byte)'a' and <= (byte)'f' => b - (byte)'a' + 10,
        _ => -1,
    };
}
